### ISSUE TITLE
`trigger_files` doesn't seem to work when a PR is opened

### ISSUE THREAD

**Issue by camelid** [rust-lang/triagebot#1535]
For example: https://github.com/rust-lang/rust/pull/91833

That PR was force-pushed here: https://github.com/rust-lang/rust/pull/91833#event-5753091480

Both the old and the new versions modified rustdoc, but only after the force-push did the PR get autolabeled:

![image](https://user-images.githubusercontent.com/37223377/145725845-9a95cd85-b022-44da-b5e5-69bf626ec65f.png)

@Mark-Simulacrum I suspect there's something wrong in `diff_between()`. For one, it uses `unwrap_or_default` for the commit hashes, which seems like it'd introduce weird behavior. If the value isn't there, it'll just use an empty commit hash.

When you get a chance, could you look at the logs for that PR perhaps? Other than that, I don't know how to fix this.

**Comment by camelid** [comment]
Actually, autolabeling just seems broken altogether; it just added `T-rustdoc` to #91819.

**Comment by klensy** [comment]
This https://github.com/rust-lang/rust/pull/91825#issuecomment-991880400 can be related?

**Comment by camelid** [comment]
No, that was fixed already.

**Comment by Mark-Simulacrum** [comment]
Okay, after some trial and error I think this is fixed on the current deploy; tests with force-pushes worked here https://github.com/rust-lang/rust/pull/72571#event-5753397775

I haven't tested opening a new PR yet -- but I think we can move ahead on rust-lang/rust#91819 at least.

**Comment by camelid** [comment]
PR approved :)

**Comment by camelid** [comment]
I just realized a potential issue that may come up with this feature: If rustbot incorrectly labels a PR with, e.g., `T-compiler`, then every time the PR is pushed to, the label will have to be manually removed. I imagine it won't be an issue too often in practice though.

**Comment by Mark-Simulacrum** [comment]
Without storing metadata in the database I'm not sure how much better we can do. It'd theoretically be pretty cheap to add the database state but I'm a little reluctant to introduce that complexity.

That said it would be relatively simple to check just the pre-sync commit, I suppose, to see if that was already enough that the label should have been added..

In general it doesn't seem worth it to me though.

**Comment by camelid** [comment]
Yeah, I think we should only try to "fix" it if we run into issues.

**Comment by camelid** [comment]
> I haven't tested opening a new PR yet

Ok, I just opened [a new PR](https://github.com/rust-lang/rust/pull/91846) and the autolabeling didn't work. So I think there's still a bug in that part of the code.

Also, [highfive thinks I'm a new contributor](https://github.com/rust-lang/rust/pull/91846#issuecomment-992038870) for some reason?

**Comment by Mark-Simulacrum** [comment]
> Also, highfive thinks I'm a new contributor for some reason?

IIRC, this is based on the last ~100 PRs or something like that? Not sure.

> Ok, I just opened a new PR and the autolabeling didn't work. So I think there's still a bug in that part of the code.

Yeah, seems so. Looking at that particular event, it looks like we .. just didn't deserialize the pull_request field. I think that's another case of serde not having compile-time duplicate detection, pull_request was already alias'd to the issue field. Pushing a fix shortly.

**Comment by camelid** [comment]
Fix confirmed to work: #91876

**Comment by Mark-Simulacrum** [comment]
https://github.com/rust-lang/rust/pull/89831#event-5760128523 has T-rustdoc being added spuriously(?)

**Comment by Mark-Simulacrum** [comment]
That generated an event with before/after commits of rust-lang/rust@62e7bcd41fed2bd7f29e66498d51d5519f12ca74 and rust-lang/rust@314027bb8fe084f7a21268dc743505c0f4f5b2c5, which had an intermediate rebase over a bunch of commits.

Direct comparison https://github.com/rust-lang/rust/compare/62e7bcd41fed2bd7f29e66498d51d5519f12ca74...314027bb8fe084f7a21268dc743505c0f4f5b2c5 produces a diff that contains rustdoc files, so the bug lies in comparing the two. Not sure if GitHub supports a merge-base comparison...

**Comment by Mark-Simulacrum** [comment]
I suspect the immediate fix here is to use base/head always, not try to only apply labels on the just-pushed changes. It does mean we get more accidental labeling per https://github.com/rust-lang/triagebot/issues/1535#issuecomment-992000852, but my guess is not *significantly* more, so I'm not too worried by it. Pushed a commit up to https://github.com/rust-lang/triagebot/pull/1540 to do this.
